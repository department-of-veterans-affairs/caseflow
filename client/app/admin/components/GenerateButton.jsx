/* eslint-disable func-style */
import React, { useEffect, useState } from 'react';
import Button from '../../components/Button';
import Modal from '../../components/Modal';
import Alert from '../../components/Alert';
import LoadingContainer from '../../components/LoadingContainer';
import { LOGO_COLORS } from 'app/constants/AppConstants';
import PropTypes from 'prop-types';
import { CSVLink } from 'react-csv';

const GenerateButton = (props) => {

  // state properties
  const [showConfirmModal, setShowConfirmModal] = useState(false);
  const [isLoading, setIsLoading] = useState(false);
  const [showEmptyResultsBanner, setShowEmptyResultsBanner] = useState(false);
  const [showErrorBanner, setShowErrorBanner] = useState(false);
  const [extractedResults, setExtractedResults] = useState('');

  useEffect(() => {
    setIsLoading(props.isLoading);
  }, [props.isLoading]);

  useEffect(() => {
    setExtractedResults(props.extractedResults);

    if (props.extractedResults?.length > 0) {
      setShowEmptyResultsBanner(false);
      setShowConfirmModal(true);
    }
  }, [props.extractedResults]);

  useEffect(() => {
    if (props.emptyResultsMessage?.length > 0) {
      setShowEmptyResultsBanner(true);
      setShowConfirmModal(false);
    }
  }, [props.emptyResultsMessage]);

  useEffect(() => {
    if (props.manualExtractionSuccess === false) {
      setShowErrorBanner(true);
    } else {
      setShowErrorBanner(false);
    }
  }, [props.manualExtractionSuccess]);

  const onClickGenerate = () => {
    props.sendExtractRequest();
  };

  return (
    <div style={{ height: '75vh' }}>
      {
        showEmptyResultsBanner &&
        <div style={{ padding: '10px' }}>
          <Alert message="No Veterans were found" type="success" />
        </div>
      }
      {
        showErrorBanner &&
        <div style={{ padding: '10px' }}>
          <Alert message="Veteran Extract Failed" type="error" />
        </div>
      }
      {
        !isLoading &&
        <Button
          id="generate-extract"
          onClick={() => onClickGenerate()}
        >
          Generate Veteran Extract
        </Button>
      }
      {
        showConfirmModal &&

        <Modal title="This file contains PII"
          confirmButton={<CSVLink data={extractedResults} filename="veteran_extract.csv" onClick={() => setShowConfirmModal(false)}>Confirm</CSVLink>}
          closeHandler={() => {
            setShowConfirmModal(false);
          }}
        >
          Are you sure you want to download it?
        </Modal>
      }
      {
        isLoading &&

        <LoadingContainer color={LOGO_COLORS.QUEUE.ACCENT}>
          <div className="loading-div">
            Action is Running...
          </div>
        </LoadingContainer>
      }
    </div>
  );
};

GenerateButton.propTypes = {
  extractedResults: PropTypes.string,
  manualExtractionSuccess: PropTypes.bool,
  isLoading: PropTypes.bool,
  sendExtractRequest: PropTypes.func,
  emptyResultsMessage: PropTypes.string,
};

export default GenerateButton;
