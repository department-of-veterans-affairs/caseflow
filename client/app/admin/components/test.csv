VET_LAST_NAME,VET_FIRST_NAME,VET_MIDDLE_NAME,VET_DATE_OF_BIRTH,VET_SSN,VET_PARTICIPANT_ID,VET_FILE_NUMBER,APPELLANT_LAST_NAME,APPELLANT_FIRST_NAME,APPELLANT_MIDDLE_NAME,APPELLANT_DATE_OF_BIRTH,APPELLANT_SSN,APPELLANT_GENDER,APPELLANT_ADDRESS,APPELLANT_PHONE,APPELLANT_EMAIL,APPELLANT_EDI_PI,APPELLANT_CORP_PID,APPELLANT_VACOLS_INTERNAL_ID,RELATIONSHIP_TO_VETERAN