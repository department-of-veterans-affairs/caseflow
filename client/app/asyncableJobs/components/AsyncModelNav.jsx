import React from 'react';
import moment from 'moment';
import PropTypes from 'prop-types';
import DropdownButton from '../../components/DropdownButton';

const DATE_TIME_FORMAT = 'ddd MMM DD HH:mm:ss YYYY';

export default class AsyncModelNav extends React.PureComponent {
  modelNameLinks = () => {
    const links = [];
    const modelNames = this.props.models.sort();
    const numLinks = modelNames.length;

    for (let modelName of modelNames) {
      const url = `/asyncable_jobs/${modelName}/jobs`;
      let modelLink;

      if (numLinks > 4) {
        modelLink = {
          title: modelName,
          target: url
        };
      } else {
        modelLink = <span key={modelName} className="cf-model-jobs-link"><a href={url}>{modelName}</a></span>;
      }

      links.push(modelLink);
    }

    return numLinks > 4 ? <DropdownButton lists={links} label="Filter by Job Type" /> : links;
  }

  render = () => {
    const currentPath = this.props.asyncableJobKlass ?
      `/asyncable_jobs/${this.props.asyncableJobKlass}/jobs` :
      '/jobs';

    return <div>
      <strong>Last updated:</strong> {moment(this.props.fetchedAt).format(DATE_TIME_FORMAT)}
      <div style={{ marginTop: '.5em' }}>
        <a style={{ marginRight: '.5em' }} href="/jobs" className="cf-link-btn">All jobs</a>
        {this.modelNameLinks()}
        <a style={{ float: 'right' }} href={`${currentPath}.csv`} className="cf-link-btn">Download as CSV</a>
      </div>
    </div>;
  }
}

AsyncModelNav.propTypes = {
  models: PropTypes.array,
  fetchedAt: PropTypes.string,
  asyncableJobKlass: PropTypes.string
};
