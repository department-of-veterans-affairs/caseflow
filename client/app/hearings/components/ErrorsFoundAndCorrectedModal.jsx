import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';

import Modal from '../../components/Modal';
import Button from '../../components/Button';
import FileUpload from '../../components/FileUpload';
import TextareaField from '../../components/TextareaField';

const ErrorsFoundAndCorrectedModal = (props) => {
  const [isSubmitDisabled, setIsSubmitDisabled] = useState(true);
  const [loading, setLoading] = useState(false);
  const [isAnyFileSelected, setIsAnyFileSelected] = useState(false);
  const [selectedFile, setSelectedFile] = useState({});
  const [fileInputContainerClassName, setFileInputContainerClassName] = useState('cf-file-input-container');
  const [notes, setNotes] = useState('');

  useEffect(() => {
    if (loading) {
      setIsSubmitDisabled(true);
    } else if (isAnyFileSelected && notes.trim() !== '') {
      setIsSubmitDisabled(false);
    } else if (notes === '') {
      setIsSubmitDisabled(true);
    }
  }, [loading, isAnyFileSelected, notes]);

  const cancel = () => {
    props.closeModal();
  };

  const submit = () => {
    // setLoading(true);
    console.log(selectedFile);
    console.log(notes);
  };

  const handleFileChange = (file) => {
    setIsAnyFileSelected(true);
    setSelectedFile(file);
    setFileInputContainerClassName('cf-file-input-container-selected');
  };

  const handleTextareaFieldChange = (event) => {
    setNotes(event);
  };

  return (
    <Modal
      title="Upload transcript to VBMS"
      closeHandler={cancel}
      confirmButton={<Button disabled={isSubmitDisabled} onClick={submit}>Upload to VBMS</Button>}
      cancelButton={<Button disabled={loading} linkStyling onClick={cancel}>Cancel</Button>}
    >
      <p>Please upload the revised transcript file for upload to VBMS.</p>
      <strong style={{ color: 'black' }}>Please select PDF</strong>
      {!isAnyFileSelected &&
        <div className={fileInputContainerClassName}>
          <FileUpload
            preUploadText="Choose from folder"
            id="cf-file-input"
            fileType=".pdf"
            onChange={handleFileChange}
          />
        </div>
      }
      {isAnyFileSelected &&
        <div className="cf-file-container-selected-container">
          <div className={fileInputContainerClassName} style={{ borderBottom: '1px solid white' }}>
            <div className="cf-file-input-row-element">
              <strong style={{ color: 'black' }}>Selected file</strong>
            </div>
            <div className="cf-file-input-row-element">
              <u><FileUpload
                preUploadText="Change file"
                postUploadText="Change file"
                id="usa-file-input"
                fileType=".pdf"
                onChange={handleFileChange}
              /></u>
            </div>
          </div>
          <div className={fileInputContainerClassName} style={{ borderTop: '1px solid white' }}>
            { selectedFile.fileName }
          </div>
        </div>
      }
      <div className="comment-size-container">
        <TextareaField
          id="cf-form-textarea"
          name="Please provide context and instructions for this action"
          onChange={handleTextareaFieldChange}
          value={notes}
        />
      </div>
    </Modal>
  );
};

ErrorsFoundAndCorrectedModal.propTypes = {
  closeModal: PropTypes.func
};

export default ErrorsFoundAndCorrectedModal;

