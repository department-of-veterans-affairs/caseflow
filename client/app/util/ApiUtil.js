import request from 'superagent';
import nocache from 'superagent-no-cache';
import ReactOnRails from 'react-on-rails';

// TODO(jd): Fill in other HTTP methods as needed
const ApiUtil = {

  // Converts regular language to camelCase
  // 'VACOLS - 123, User' becomes 'vacolsUser'
  convertToCamelCase(phrase = '') {
    // Code courtesy of Stack Overflow, Question 2970525
    return phrase.toLowerCase().
        replace(/[^a-zA-Z ]/g, "").
        replace(/(?:^\w|[A-Z]|\b\w|\s+)/g, (match, index) => {
          if (Number(match) === 0) {
            return "";
          }

          return index === 0 ? match.toLowerCase() : match.toUpperCase();
        });
  },

  // Converts camelCase to snake_case
  convertToSnakeCase(data = {}) {
    let result = {};

    for (let key in data) {
      if ({}.hasOwnProperty.call(data, key)) {
        // convert key from camelCase to snake_case
        let snakeKey = key.replace(/([A-Z])/g, ($1) => `_${$1.toLowerCase()}`);

        // assign value to new object
        result[snakeKey] = data[key];
      }
    }

    return result;
  },

  get(url, options = {}) {
    return request.
      get(url).
      set(this.headers(options.headers)).
      query(options.query).
      use(nocache);
  },

  // Default headers needed to talk with rails server.
  // Including rail's authenticity token
  headers(options = {}) {
    let headers = Object.assign({},
      {
        'Accept': 'application/json',
        'Content-Type': 'application/json'
      },
      ReactOnRails.authenticityHeaders(),
      options);


    return headers;
  },

  patch(url, options = {}) {
    return request.
      post(url).
      set(this.headers({ 'X-HTTP-METHOD-OVERRIDE': 'patch' })).
      send(options.data).
      use(nocache);
  },

  post(url, options = {}) {
    return request.
      post(url).
      set(this.headers(options.headers)).
      send(options.data).
      use(nocache);
  }
};

export default ApiUtil;
