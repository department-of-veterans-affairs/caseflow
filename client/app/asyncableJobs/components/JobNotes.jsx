import React from 'react';
import Button from '../../components/Button';
import ApiUtil from '../../util/ApiUtil';
import PropTypes from 'prop-types';
import moment from 'moment';
import ReactMarkdown from 'react-markdown';

const DATE_TIME_FORMAT = 'ddd MMM DD HH:mm:ss YYYY';

import Checkbox from '../../components/Checkbox';
import TextareaField from '../../components/TextareaField';

class NewNoteForm extends React.PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      saveInProgress: false,
      checkboxSelected: true,
      newNote: ''
    };
  }

  onSave = () => {
    const job = this.props.job;
    const url = `/asyncable_jobs/${job.klass}/jobs/${job.id}/note`;
    const form = this;
    const notes = this.props.notes;

    const data = {
      note: this.state.newNote,
      send_to_intake_user: this.state.checkboxSelected
    };

    notes.setState({ reloaded: false });

    ApiUtil.post(url, { data }).
      then(
        (response) => {
          const newNote = JSON.parse(response.text);

          form.setState({
            saveInProgress: false,
            checkboxSelected: false,
            newNote: ''
          });

          // add note to current page.
          notes.props.notes.unshift(newNote);
          notes.setState({ reloaded: true });
        }
      );
  }

  onChange = (event) => {
    this.setState({ newNote: event });
  }

  onCheckboxChange = (event) => {
    this.setState({ checkboxSelected: event });
  }

  isSaveDisabled = () => {
    return this.state.saveInProgress || this.state.newNote.trim().length === 0;
  }

  isCheckboxChecked = () => {
    return this.state.checkboxSelected;
  }

  sendToUserCheckbox = () => {
    return <Checkbox
      label="Send as message to user"
      name="send_to_intake_user"
      value={this.isCheckboxChecked()}
      onChange={this.onCheckboxChange}
      disabled={this.state.saveInProgress}
    />;
  }

  render = () => {
    return <div className="comment-size-container">
      <TextareaField
        hideLabel
        name="Add Note"
        aria-label="Add Note"
        onChange={this.onChange}
        value={this.state.newNote}
      />
      <div className="comment-save-button-container">
        <span className="cf-right-side">
          { this.props.job.user && this.sendToUserCheckbox() }
          <Button
            name="save"
            disabled={this.isSaveDisabled()}
            onClick={this.onSave}>
              Add Note
          </Button>
        </span>
      </div>
    </div>;
  }

}

NewNoteForm.propTypes = {
  job: PropTypes.object,
  notes: PropTypes.object,
  onSave: PropTypes.func,
  onCancel: PropTypes.func
};

export default class JobNotes extends React.PureComponent {

  render = () => {
    const { notes, job } = this.props;

    return <div className="job-notes">
      <h3>Notes</h3>
      <NewNoteForm job={job} notes={this} />
      <div>
        { notes.map((note, index) => {
          return <div className="job-note" key={`job-note-container-${index}`}>
            <div
              className="job-note-details"
              data-key={`job-note-${index}`}
              key={`job-note-${index}`}
              id={`job-note-${note.id}`}>

              <div>
                <span className="job-note-time">{moment(note.created_at).format(DATE_TIME_FORMAT)}</span>
                <span className="job-note-user">{note.user}</span>
              </div>
              <div className="job-note-note"><ReactMarkdown source={note.note} /></div>

            </div>
          </div>;
        })}
      </div>
    </div>;
  }
}

JobNotes.propTypes = {
  job: PropTypes.object,
  notes: PropTypes.array
};
