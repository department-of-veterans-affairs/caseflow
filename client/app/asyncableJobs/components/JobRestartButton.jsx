import React from 'react';
import Button from '../../components/Button';
import ApiUtil from '../../util/ApiUtil';
import PropTypes from 'prop-types';

class JobRestartButton extends React.PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      restarted: false,
      restarting: false
    };
  }

  restartJob = () => {
    this.setState({ restarting: true });
    this.sendRestart();
  }

  sendRestart = () => {
    const button = this;
    const page = this.props.page;
    const job = this.props.job;
    const url = `/asyncable_jobs/${job.klass}/jobs/${job.id}`;

    ApiUtil.patch(url, {}).
      then(
        (response) => {
          const responseObject = JSON.parse(response.text);

          Object.assign(job, responseObject);

          job.error = '';
          job.restarted = true;

          button.setState({
            restarted: true,
            restarting: false
          });

          // trigger table rerender
          const jobsRestarted = page.state.restarted + 1;

          page.setState({ restarted: jobsRestarted });
        }
      );
  }

  getButtonClassNames = () => {
    let classNames = ['usa-button'];

    if (this.state.restarting || this.state.restarted) {
      classNames.push('usa-button-disabled');
    }

    return classNames;
  }

  getButtonText = () => {
    let txt = 'Restart';

    if (this.state.restarting) {
      txt = 'Restarting';
    } else if (this.state.restarted) {
      txt = 'Restarted';
    } else if (this.props.job.processed_at) {
      txt = 'Processed';
    } else if (this.disableRestart()) {
      txt = 'Queued';
    }

    return txt;
  }

  disableRestart = () => {
    const job = this.props.job;

    if (job.processed_at) {
      return true;
    }

    if (!job.attempted_at) {
      return true;
    }

    const fiveMinutes = 300000;

    const lastAttempted = new Date(job.attempted_at).getTime();
    const submittedAt = new Date(job.last_submitted_at).getTime();
    const now = new Date().getTime();

    if ((now - lastAttempted) < fiveMinutes || (now - submittedAt) < fiveMinutes) {
      return true;
    }

    return false;
  }

  render = () => {
    const job = this.props.job;

    return <Button
      id={`job-${job.klass}-${job.id}`}
      title={`${job.klass} ${job.id}`}
      loading={this.state.restarting}
      loadingText="Restarting..."
      disabled={this.disableRestart()}
      onClick={() => {
        this.restartJob();
      }}
      classNames={this.getButtonClassNames()}
    >{this.getButtonText()}</Button>;
  }
}

JobRestartButton.propTypes = {
  job: PropTypes.object,
  page: PropTypes.object
};

export default JobRestartButton;
