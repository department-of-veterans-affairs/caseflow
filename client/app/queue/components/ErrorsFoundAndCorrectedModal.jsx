import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';

import Modal from '../../components/Modal';
import Button from '../../components/Button';
import FileUpload from '../../components/FileUpload';
import TextareaField from '../../components/TextareaField';
import COPY from '../../../COPY';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { requestPatch } from '../uiReducer/uiActions';
import { withRouter } from 'react-router-dom';
import { setAppealAttrs } from '../QueueActions';

import {
  appealWithDetailSelector,
  taskById
} from '../selectors';
import ApiUtil from '../../util/ApiUtil';

const SelectedFileSection = (props) => {
  const { fileInputContainerClassName, handleFileChange, selectedFile } = props;

  return (
    <div className="cf-file-container-selected-container" style={{ marginTop: '2rem', marginBottom: '2rem' }}>
      <div className={fileInputContainerClassName} style={{ borderBottom: '1px solid white' }}>
        <div className="cf-file-input-row-element">
          <strong style={{ color: 'black' }}>Selected file</strong>
        </div>
        <div className="cf-file-input-row-element">
          <u><FileUpload
            preUploadText="Change file"
            postUploadText="Change file"
            id="usa-file-input"
            fileType=".pdf"
            onChange={handleFileChange}
          /></u>
        </div>
      </div>
      <div className={fileInputContainerClassName} style={{ borderTop: '1px solid white' }}>
        { selectedFile.fileName }
      </div>
    </div>
  );
};

export const ErrorsFoundAndCorrectedModal = (props) => {
  const [isSubmitDisabled, setIsSubmitDisabled] = useState(true);
  const [loading, setLoading] = useState(false);
  const [isAnyFileSelected, setIsAnyFileSelected] = useState(false);
  const [selectedFile, setSelectedFile] = useState({});
  const [fileInputContainerClassName, setFileInputContainerClassName] = useState('cf-file-input-container');
  const [notes, setNotes] = useState('');
  const [transcriptFileName, setTranscriptFileName] = useState('');
  const [fileNameError, setFileNameError] = useState(false);

  useEffect(() => {
    ApiUtil.get(`/tasks/${props.task.taskId}/uploaded_transcription_file`, {
      query: { appeal_id: props.appealId }
    }).then((response) => {
      setTranscriptFileName(response.body.file_name);
    });
  }, []);

  useEffect(() => {
    if (loading) {
      setIsSubmitDisabled(true);
    } else if (fileNameError || notes === '') {
      setIsSubmitDisabled(true);
    } else if (isAnyFileSelected && notes.trim() !== '') {
      setIsSubmitDisabled(false);
    }
  }, [loading, isAnyFileSelected, notes, fileNameError]);

  const cancel = () => {
    props.closeModal();
  };

  const redirectAfterSubmit = () => {
    window.location = `/queue/appeals/${props.appealId}`;
  };

  const submit = () => {
    const { task } = props;

    const formatInstructions = () => {
      return [
        COPY.REVIEW_TRANSCRIPT_TASK_DEFAULT_INSTRUCTIONS,
        COPY.UPLOAD_TRANSCRIPTION_VBMS_ERRORS_ACTION_TYPE,
        notes,
        selectedFile.fileName
      ];
    };

    const requestParams = () => {
      return {
        data: {
          task: {
            instructions: formatInstructions()
          },
          file_info: {
            file: selectedFile.file,
            file_name: selectedFile.fileName
          },
          appeal_id: props.appealId
        }
      };
    };

    setLoading(true);

    return props.requestPatch(`/tasks/${task.taskId}
      /error_found_upload_transcription_to_vbms`, requestParams()
    ).then(() => {
      redirectAfterSubmit();
    });
  };

  const handleFileChange = (file) => {
    if (file.fileName === transcriptFileName) {
      setFileInputContainerClassName('cf-file-input-container-selected');
      setFileNameError(false);
    } else {
      setFileInputContainerClassName('cf-file-input-container-selected cf-file-error-container');
      setFileNameError(true);
    }
    setIsAnyFileSelected(true);
    setSelectedFile(file);
  };

  const handleTextareaFieldChange = (event) => {
    setNotes(event);
  };

  return (
    <Modal
      title="Upload transcript to VBMS"
      closeHandler={cancel}
      confirmButton={<Button disabled={isSubmitDisabled} onClick={submit}>Upload to VBMS</Button>}
      cancelButton={<Button disabled={loading} linkStyling onClick={cancel}>Cancel</Button>}
    >
      <p>Please upload the revised transcript file for upload to VBMS.</p>
      <p>
        <strong style={{ color: 'black' }}>Transcript file name</strong>
        <div>{transcriptFileName}</div>
      </p>
      <strong style={{ color: 'black' }}>Please select PDF</strong>
      {!isAnyFileSelected &&
        <div className={fileInputContainerClassName}>
          <FileUpload
            preUploadText="Choose from folder"
            id="cf-file-input"
            fileType=".pdf"
            onChange={handleFileChange}
          />
        </div>
      }
      {fileNameError &&
        <div className="usa-input-error" style={{ right: 0 }}>
          <strong style={{ color: '#cd2026'}}>
            File name must exactly match the transcript file name. Ensure there are no spaces or extra characters.
          </strong>
          <SelectedFileSection
            handleFileChange={handleFileChange}
            fileInputContainerClassName={fileInputContainerClassName}
            selectedFile={selectedFile}
          />
        </div>
      }
      {!fileNameError && isAnyFileSelected &&
        <SelectedFileSection
          handleFileChange={handleFileChange}
          fileInputContainerClassName={fileInputContainerClassName}
          selectedFile={selectedFile}
        />
      }
      <div className="comment-size-container">
        <TextareaField
          id="cf-form-textarea"
          name="Please provide context and instructions for this action"
          onChange={handleTextareaFieldChange}
          value={notes}
        />
      </div>
    </Modal>
  );
};

ErrorsFoundAndCorrectedModal.propTypes = {
  closeModal: PropTypes.func,
  task: PropTypes.shape({
    taskId: PropTypes.string,
    type: PropTypes.string,
  }),
  requestPatch: PropTypes.func,
  appeal: PropTypes.shape({
    veteranFullName: PropTypes.string
  }),
  appealId: PropTypes.string,
  error: PropTypes.shape({
    title: PropTypes.string,
    detail: PropTypes.string
  }),
};

SelectedFileSection.propTypes = {
  handleFileChange: PropTypes.func,
  fileInputContainerClassName: PropTypes.string,
  selectedFile: PropTypes.object
};

const mapStateToProps = (state, ownProps) => ({
  error: state.ui.messages.error,
  appeal: appealWithDetailSelector(state, ownProps),
  task: taskById(state, { taskId: ownProps.taskId })
});

const mapDispatchToProps = (dispatch) => bindActionCreators({
  requestPatch,
  setAppealAttrs
}, dispatch);

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(ErrorsFoundAndCorrectedModal));

