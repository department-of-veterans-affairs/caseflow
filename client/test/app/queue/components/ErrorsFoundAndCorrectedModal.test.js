import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { axe } from 'jest-axe';

import { ErrorsFoundAndCorrectedModal } from '../../../../app/queue/components/ErrorsFoundAndCorrectedModal';
import ApiUtil from '../../../../app/util/ApiUtil';

describe('ErrorsFoundAndCorrectedModal', () => {
  const closeModal = jest.fn();
  const requestPatch = jest.fn();
  const getSpy = jest.spyOn(ApiUtil, 'get');

  const defaultProps = {
    taskId: '1000',
    closeModal,
    requestPatch,
    type: 'ReviewTranscriptTask',
    appealId: '12',
    task: {
      taskId: '1000',
      type: 'ReviewTranscriptTask'
    }

  };

  beforeEach(() => {
    getSpy.mockImplementation(() => new Promise((resolve) => resolve({ body: { file_name: 'test.pdf' } })));
  });

  it('renders correctly', () => {

    const { container } = render(<ErrorsFoundAndCorrectedModal {...defaultProps} />);

    expect(container).toMatchSnapshot();
  });

  it('passes a11y testing', async () => {
    const { container } = render(<ErrorsFoundAndCorrectedModal {...defaultProps} />);

    const results = await axe(container);

    expect(results).toHaveNoViolations();
  });

  it('displays the default page elements with default props', () => {
    render(
      <ErrorsFoundAndCorrectedModal {...defaultProps} />
    );

    const textarea = screen.getByRole('textbox');

    expect(screen.getByText('Upload transcript to VBMS')).toBeInTheDocument();
    expect(screen.getByText('Please upload the revised transcript file for upload to VBMS.')).
      toBeInTheDocument();
    expect(screen.getByText('Please select PDF')).toBeInTheDocument();
    expect(screen.getByText('Choose file from folder')).toBeInTheDocument();
    expect(screen.getByText('Please provide context and instructions for this action')).
      toBeInTheDocument();
    expect(textarea.value).toBe('');
  });

  it('can handle a file upload', async () => {
    const file = new File(['test'], 'test.pdf', { type: 'application/pdf' });

    render(
      <ErrorsFoundAndCorrectedModal {...defaultProps} />
    );

    const input = screen.getByLabelText(/Choose file from folder/i);

    userEvent.upload(input, file);

    await waitFor(() => {
      expect(input.files.length).toBe(1);
      // Two instances of file names. Existing file name and file name uploaded by user
      expect(screen.getAllByText('test.pdf')).toHaveLength(2);
      expect(screen.getByText('Selected file')).toBeInTheDocument();
      expect(screen.getByText('Change file')).toBeInTheDocument();
    });
  });

  it('display error when file name does not match the uploaded file name', async () => {
    const file = new File(['test1'], 'test1.pdf', { type: 'application/pdf' });

    render(
      <ErrorsFoundAndCorrectedModal {...defaultProps} />
    );

    expect(screen.getByText('Upload to VBMS').closest('button')).
      toBeDisabled();
    expect(
      screen.queryByText(/File name must exactly macth the transcript file name/i)
    ).not.toBeInTheDocument();

    const input = screen.getByLabelText(/Choose file from folder/i);
    const textarea = screen.getByRole('textbox');

    userEvent.upload(input, file);

    userEvent.type(textarea, 'This is a note.');

    await waitFor(() => {
      expect(
        screen.getByText(/File name must exactly match the transcript file name/i)
      ).toBeInTheDocument();
      expect(screen.getByText('Upload to VBMS').closest('button')).
        toBeDisabled();
    });
  });

  it('the submit button is enabled when file name matched and fields filled out', async () => {
    const file = new File(['test'], 'test.pdf', { type: 'application/pdf' });

    render(
      <ErrorsFoundAndCorrectedModal {...defaultProps} />
    );

    expect(screen.getByText('Upload to VBMS').closest('button')).
      toBeDisabled();

    const input = screen.getByLabelText(/Choose file from folder/i);
    const textarea = screen.getByRole('textbox');

    userEvent.upload(input, file);

    await waitFor(() => {
      expect(screen.getByText('Upload to VBMS').closest('button')).
        toBeDisabled();
    });

    userEvent.type(textarea, 'This is a note.');

    await waitFor(() => {
      expect(screen.getByText('Upload to VBMS').closest('button')).
        toBeEnabled();
    });
  });
});
