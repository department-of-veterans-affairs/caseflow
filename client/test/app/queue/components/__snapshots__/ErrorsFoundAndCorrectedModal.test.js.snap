// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorsFoundAndCorrectedModal renders correctly 1`] = `
<div>
  <section
    aria-describedby="modal_id-desc"
    aria-labelledby="modal_id-title"
    aria-modal="true"
    class="cf-modal active "
    id="modal_id"
    role="dialog"
  >
    <div
      class="cf-modal-body"
      id=""
    >
      <button
        class="cf-modal-close"
        id="Upload-transcript-to-VBMS-button-id-close"
        type="button"
      >
        <span
          class="usa-sr-only"
        >
          Close
        </span>
        <svg
          class="cf-icon-close"
          height="55"
          viewBox="0 0 55 55"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title>
            close
          </title>
          <g
            fill="#212121"
          >
            <path
              d="M52.6 46.9l-6 6c-.8.8-1.9 1.2-3 1.2s-2.2-.4-3-1.2l-13-13-13 13c-.8.8-1.9 1.2-3 1.2s-2.2-.4-3-1.2l-6-6c-.8-.8-1.2-1.9-1.2-3s.4-2.2 1.2-3l13-13-13-13c-.8-.8-1.2-1.9-1.2-3s.4-2.2 1.2-3l6-6c.8-.8 1.9-1.2 3-1.2s2.2.4 3 1.2l13 13 13-13c.8-.8 1.9-1.2 3-1.2s2.2.4 3 1.2l6 6c.8.8 1.2 1.9 1.2 3s-.4 2.2-1.2 3l-13 13 13 13c.8.8 1.2 1.9 1.2 3s-.4 2.2-1.2 3z"
            />
          </g>
        </svg>
      </button>
      <div
        style="display: flex; flex-wrap: wrap;"
      >
        <h1
          id="modal_id-title"
          style="flex: 1;"
        >
          Upload transcript to VBMS
        </h1>
        <div
          data-css-2flkvz=""
        >
          <p>
            Please upload the revised transcript file for upload to VBMS.
          </p>
          <strong
            style="color: black;"
          >
            Please select PDF
          </strong>
          <div
            class="cf-file-input-container"
          >
            <div>
              <label
                for="cf-file-input"
              >
                <a>
                  Choose from folder
                </a>
              </label>
              <div
                data-css-gyrv08=""
              >
                <input
                  accept=".pdf"
                  id="cf-file-input"
                  type="file"
                />
              </div>
            </div>
          </div>
          <div
            class="comment-size-container"
          >
            <div
              class="cf-form-textarea"
            >
              <label
                class="question-label"
                for="cf-form-textarea"
              >
                <span>
                  Please provide context and instructions for this action
                </span>
              </label>
              <textarea
                id="cf-form-textarea"
                name="Please provide context and instructions for this action"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="cf-modal-divider"
      />
      <div
        class="cf-modal-controls"
      >
        <div>
          <span
            class="cf-push-right"
          >
            <span>
              <button
                class="cf-submit usa-button-disabled usa-button"
                disabled=""
                type="button"
              >
                Upload to VBMS
              </button>
            </span>
          </span>
          <span
            class="cf-push-left"
          >
            <span>
              <button
                class="cf-submit cf-btn-link usa-button"
                type="button"
              >
                Cancel
              </button>
            </span>
          </span>
        </div>
      </div>
    </div>
  </section>
</div>
`;
